src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs

```diff
@@ -127,8 +127,14 @@ public HostBase(HostEnvironmentBase<TEnv> source, string shortName, string paren
                 Depth = source.Depth + 1;
             }
 
+            /// <summary>
+            /// This method registers and returns the host for the calling component. The generated host is also
+            /// added to <see cref="_children"/> and encapsulated by <see cref="WeakReference"/>. It becomes
+            /// necessary to remove these hosts when they are reclaimed by the Garbage Collector.
+            /// </summary>
             public new IHost Register(string name, int? seed = null, bool? verbose = null)
             {
+                _children.RemoveAll(r => r.TryGetTarget(out IHost _) == false);
```

Comment:
I wonder: is there a reason to add this only in here on `HostEnvironmentBase<TEnv>.HostBase` but not on the method it's shadowing on `HostEnvironmentBase<TEnv>.Register`??

Reply:
Are we going to hit multi-threading issues at all? You can't iterate over a list (which is what RemoveAll is doing) and add to the list from 2 different threads at the same time. You need to synchronize the accesses of the list so they don't collide.

Reply:
@antoniovs1029 I added it only in the shadowed `HostEnvironmentBase<TEnv>.HostBase.Register()` so that `RemoveAll()` could be called fewer times, as I noticed calls to both `HostEnvironmentBase<TEnv>.HostBase.Register()` and `HostEnvironmentBase<TEnv>.Register()` occur from transformers. Other than this there is no other reason to not add it in `HostEnvironmentBase<TEnv>.Register()` too.